app/controllers/application_controller.rb

```diff
@@ -59,30 +61,26 @@ def valid_help_actions
     @valid_help_actions ||= %w[signed_in create].freeze
   end
 
-  def confirm_two_factor_setup
-    user = request.env["warden"].user
-
-    if Rails.configuration.enable_enhanced_2fa_experience
-      return if user.nil?
-
-      if user.second_factor_method.present?
-        # Handled by Devise's controller.
-        return if user.second_factor_method == "app"
+  def tfa_feature_flag?
+    Rails.configuration.enable_enhanced_2fa_experience
+  end
 
-        # Handled by us.
-        redirect_to users_two_factor_authentication_email_path
-        return
-      end
+  def choose_two_factor_method
+    return if user.nil? || !user.second_factor_method.nil?
 
-      # TODO: Replace with route helper once fully migrated to enhanced 2FA.
-      redirect_to "/users/two_factor_authentication/setup"
-      return
-    end
+    # TODO: Replace with route helper once fully migrated to enhanced 2FA.
+    redirect_to "/users/two_factor_authentication/setup"
+  end
 
+  def confirm_two_factor_setup
     return unless user &&
       user.need_two_factor_authentication?(request) &&
       !user.totp_enabled?
 
     redirect_to users_two_factor_authentication_setup_path
   end
+
+  def user
```

Comment:
shouldn't this be `current_user` throughout?

Reply:
which is provided out of the box by Devise

Reply:
Probably. Copy-pasted from what was there. I'll change it